pkg/controllers/state/cluster.go

```diff
@@ -173,9 +173,13 @@ func (c *Cluster) Nodes() StateNodes {
 	c.mu.RLock()
 	defer c.mu.RUnlock()
 
-	return lo.Map(lo.Values(c.nodes), func(n *StateNode, _ int) *StateNode {
-		return n.DeepCopy()
-	})
+	return StateNodes{
+		Nodes: lo.Map(lo.Values(c.nodes), func(n *StateNode, _ int) *StateNode {
+			return n.DeepCopy()
+		}),
+		kubeClient: c.kubeClient,
```

Comment:
I was wondering if kubeClient should live in state node itself.

Reply:
> I was wondering if kubeClient should live in state node itself.

I'm still really strongly against this pattern of storing kubeClients in data objects. If we need some kind of data accessor, we can build a separate client for that and instantiate it, but storing that kubeClient many times across all StateNodes feels wrong to me

Reply:
I'd be more in favor of something like `state.NewClient(ctx, kubeClient)` or similar that allows us to build the client from whatever thing we are injecting our dependencies in from and then building methods off of that.

Reply:
> I'm still really strongly against this pattern of storing kubeClients in data objects

Actually, after thinking about this a bit deeper, I may be completely eating my own words here 😆 . I'm coming around to the idea and think the pattern may actually be reasonable to think about conceptually, "The StateNode stores state about a NodeClaim/Node and has mechanism to grab pods directly from the cluster (because the StateNode doesn't directly store pods from informers)." I'm fine either way. Definitely worth a try if we think that it's going to improve the code readabilitiy

Reply:
> storing kubeClients in data objects

This is an old idea from the RAII pattern from C++